### PR TITLE
QR-generator project is ready for test

### PR DESCRIPTION
**Pull request by tripleCC**
I finished the test project with all requirements just now.  
The README file which can help you to run the server and application is in the root directory of my project.
Looking forward to your reply.
Thanks.

### CODE REVIEW

File: QR-generator/QRGenerator/QRGenerator/Util/Archivable.swift [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+//
+//  Created by tripleCC on 2018/1/24.
+//  Copyright © 2018年 tripleCC. All rights reserved.
+//
+
+import Foundation
+
+public protocol Archivable {
+    func archive() -> [String : Any]
+    init?(unarchive: [String : Any]?)
```
Comment (WXGBridgeQ) [inline QR-generator/QRGenerator/QRGenerator/Util/Archivable.swift]: Failable Initializer with an optional param is not a good idea, should change to non-optional value

File: QR-generator/QRGenerator/QRGenerator/Model/Seed.swift [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+//  Created by tripleCC on 2018/1/23.
+//  Copyright © 2018年 tripleCC. All rights reserved.
+//
+
+import Foundation
+import ObjectMapper
+
+struct Seed: Mappable {
+    var seed: String?
+    var expiresAt: String?
```
Comment (WXGBridgeQ) [inline QR-generator/QRGenerator/QRGenerator/Model/Seed.swift:14]: Using Date here maybe better, and in `viewModel` you can use a String

File: QR-generator/QRGenerator/QRGenerator/Networking/QRGeneratorProvider.swift [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+}
+
+extension QRGenerator: TargetType {
+    public var method: Moya.Method {
+        switch self {
+        case .seed:
+            return .get
+        default:
+            return .get
+        }
```
Comment (WXGBridgeQ) [inline QR-generator/QRGenerator/QRGenerator/Networking/QRGeneratorProvider.swift:24]: Using Enum is a good idea, but don't like the case/default style here

File: QR-generator/QRGenerator/QRGenerator/Util/QRCodeGenerator.swift [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+    
+    init(data: Data) {
+        self.data = data
+    }
+    
+    public func generate() -> UIImage? {
+        let filter = CIFilter(name: "CIQRCodeGenerator")
+        filter?.setValue(data, forKey: "inputMessage")
+        filter?.setValue("Q", forKey: "inputCorrectionLevel")
+        return filter?.outputImage.flatMap { UIImage(ciImage: $0) }
```
Comment (WXGBridgeQ) [inline QR-generator/QRGenerator/QRGenerator/Util/QRCodeGenerator.swift:26]: The QR code image looks not clear, should add an optimization step here

File: QR-generator/QRGenerator/QRGenerator/ViewController/QRGeneratorViewController.swift [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+        }
+    }
+    
+    private func setupQRCodeImage(seed: String?) {
+        // generate qrcode image
+        let qrcodeImage = seed.flatMap { QRCodeGenerator(string: $0).generate() }
+        self.qrcodeImageView.image = qrcodeImage
+    }
+    
+    private func setupAutoRefreshTimer(expiresAt: String?) {
```
Comment (WXGBridgeQ) [inline QR-generator/QRGenerator/QRGenerator/ViewController/QRGeneratorViewController.swift]: Using Date here make more sense

File: QR-generator/QRGenerator/QRGeneratorTests/QRGeneratorTests.swift [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+        
+        seed.archiveToDisk()
+        unarchivedSeed = Seed.unarchiveFromDisk()
+        XCTAssertNotNil(unarchivedSeed)
+        XCTAssertEqual(seed.seed, unarchivedSeed?.seed)
+        XCTAssertEqual(seed.expiresAt, unarchivedSeed?.expiresAt)
+    }
+    
+    func testSeedAPI() {
+
```
Comment (WXGBridgeQ) [inline QR-generator/QRGenerator/QRGeneratorTests/QRGeneratorTests.swift]: You can abstract a protocol layer for Seed API, and use a Mock way to test it. Of course, you can also find another way to do it.

File: QR-generator/QRGenerator/QRGenerator/Networking/QRGeneratorProvider.swift [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+        NetworkActivityPlugin { (type, target) in
+            switch type {
+            case .began:
+                UIApplication.shared.isNetworkActivityIndicatorVisible = true
+            case .ended:
+                UIApplication.shared.isNetworkActivityIndicatorVisible = false
+            }
+        }
+    ]
+}
```
Comment (WXGBridgeQ) [inline QR-generator/QRGenerator/QRGenerator/Networking/QRGeneratorProvider.swift:106]: 👍 
good to know this plugin method, to do some extra work along with the network request

File: QR-generator/QRGenerator/QRGenerator/Model/Seed.swift (hunk 2 above)
Comment (tripleCC) [reply QR-generator/QRGenerator/QRGenerator/Model/Seed.swift:14]: I would rather let model represents real content that is returned from server.  Maybe `ViewModel` is not necessary in the situation ,  `Formatter` is more  appropriate name for the extension.

File: QR-generator/QRGenerator/QRGenerator/Networking/QRGeneratorProvider.swift (hunk 3 above)
Comment (tripleCC) [reply QR-generator/QRGenerator/QRGenerator/Networking/QRGeneratorProvider.swift:24]: Any better ideas here?

File: QR-generator/QRGenerator/QRGenerator/Model/Seed.swift (hunk 2 above)
Comment (WXGBridgeQ) [reply QR-generator/QRGenerator/QRGenerator/Model/Seed.swift:14]: @tripleCC 
The model for the app is not necessary to be same as the model for the backend, it should be easy to use in the app.

For a date param, I would like to use Date for the model and String for the ViewModel.
* String is for presenting
* Date is easy to use for other logic (for example, compare with others)

Comment (WXGBridgeQ) [reply QR-generator/QRGenerator/QRGenerator/Model/Seed.swift:14]: You can also validate the backend data format within the Initializer by using a Date parameter

**Comment by WXGBridgeQ** [comment]
Thanks for communicating and sharing opinions, we can close it now.

**Pull request closed without merging**